### PR TITLE
Entity manager + object destruction

### PR DESCRIPTION
**Pull request by SleepingSoul**

### CODE REVIEW

File: SleepEngine/Engine/Components/HierarchyNode.cpp [hunk 1]
```diff
@@ -58,17 +80,15 @@ void HierarchyNode::detachFromParent()
...
 void HierarchyNode::update(float dt)
 {
-    Base::update(dt);
-
-    for (auto& childNodeObject : m_children)
+    for (auto* const child : m_children)
     {
-        childNodeObject->update(dt);
+        child->update(dt);
     }
```
Comment (SleepingSoul) [inline SleepEngine/Engine/Components/HierarchyNode.cpp]: So HierarchyNode still updates children but doesn't own it now. EntityManager knows that it shouldn't update these objects.

File: SleepEngine/Engine/GameTemplate/Game.cpp [hunk 2]
```diff
@@ -32,66 +31,45 @@ Game::Game(size_t width, size_t height)
...
-    auto [it, wasInserted] = m_scenes.insert({ m_nextSceneID++, std::move(sceneAndIniter) });
-    assertion(wasInserted, "nextSceneID overflow");
+    auto& scene = m_scenes.emplace_back(std::move(initer), id);
 
     if (m_scenes.size() == 1)
     {
-        m_currentScene = it;
-        initCurrentScene();
+        // first scene
+        scene.init();
```
Comment (SleepingSoul) [inline SleepEngine/Engine/GameTemplate/Game.cpp]: `init` possibly can be a very long operation. Maybe we shouldn't do it automatically here?

File: SleepEngine/Engine/GameTemplate/Game.cpp [hunk 3]
```diff
@@ -32,66 +31,45 @@ Game::Game(size_t width, size_t height)
...
     }
 
-    m_currentScene->second.first.clear();
-    m_currentScene = it;
+    spdlog::get(EngineLogger)->info("Changing scene '{}' to '{}'.", m_currentScene->getID(), id);
 
-    initCurrentScene();
+    m_entityManager.clear();
+    
+    it->init();
```
Comment (SleepingSoul) [inline SleepEngine/Engine/GameTemplate/Game.cpp]: Oh shit. Found a problem here. I guess we should `forceClear` entity manager here.

Comment (SleepingSoul) [reply SleepEngine/Engine/GameTemplate/Game.cpp]: Or maybe it will be ok, should test it

File: SleepEngine/Demos/Bubbles/Entities/Bubble.cpp [hunk 4]
```diff
@@ -19,10 +19,10 @@ void Bubble::update(float dt)
 {
     auto& transform = slp::getTransform(m_object);
 
-    float const distanceFlown = static_cast <float>((transform.getPosition() - m_startPosition).length());
+    float const distanceFlown = glm::length(transform.getPosition() - m_startPosition);
```
Comment (SleepingSoul) [inline SleepEngine/Demos/Bubbles/Entities/Bubble.cpp]: vecN.length() method returns dimension of vecN, lol

File: SleepEngine/Engine/Components/HierarchyNode.cpp [hunk 5]
```diff
@@ -27,26 +43,32 @@ Object* HierarchyNode::addChild(std::unique_ptr<Object>&& child)
...
+    if (!found)
     {
-        return element.get() == child;
-    };
+        LOG_AND_FAIL("'removeChild' called for object that is not our child");
+        return;
+    }
 
-    auto const childToDelete = findIf(m_children, isChildToDelete);
+    m_children.erase(it);
```
Comment (TarasMartynyuk) [inline SleepEngine/Engine/Components/HierarchyNode.cpp]: Erase but no remove?

File: SleepEngine/Engine/EntityManager.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+BEGIN_NAMESPACE_SLEEP
+
+void EntityManager::addObject(std::unique_ptr<Object>&& object, bool shouldUpdate)
+{
+    m_addLaterObjects.emplace_back(std::move(object), shouldUpdate);
+}
+
+void EntityManager::removeObjectLater(Object const* object)
+{
+    if (!contains(m_removeLaterObjects, object))
```
Comment (TarasMartynyuk) [inline SleepEngine/Engine/EntityManager.cpp]: Assert

File: SleepEngine/Engine/GameTemplate/Game.cpp (hunk 2 above)
Comment (TarasMartynyuk) [reply SleepEngine/Engine/GameTemplate/Game.cpp]: Yeah, we should do it in game run

File: SleepEngine/Engine/EntityManager.h [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+    void update(float dt);
+
+private:
+    void postUpdate();
+
+    void addObjects();
+    void removeObjects();
+
+    EntitiesContainer m_objects;
```
Comment (TarasMartynyuk) [inline SleepEngine/Engine/EntityManager.h]: Is the alias needed? Used in 1 place, impl detal. Makes it harder to understand code.

File: SleepEngine/Engine/GameTemplate/Game.cpp (hunk 3 above)
Comment (TarasMartynyuk) [reply SleepEngine/Engine/GameTemplate/Game.cpp]: for the first time it will be ok, but only then

File: SleepEngine/Engine/GameTemplate/Scene.h [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+BEGIN_NAMESPACE_SLEEP
+
+class Scene
+{
+public:
+    using Initer = std::function<void()>;
+
+    Scene(Initer&& initer, std::string_view id);
+
+    std::string_view getID() const { return m_id; }
```
Comment (TarasMartynyuk) [inline SleepEngine/Engine/GameTemplate/Scene.h]: macros? btw i ll refact them later)

File: SleepEngine/Engine/Utils/MacroUtils.h [hunk 9]
```diff
@@ -124,7 +124,9 @@ inline componentType const& shortcutName(Object const* obj) \
 }   \
     \
 
-
+#define ALL(container) std::begin(container), std::end(container)
+#define C_ALL(container) std::cbegin(container), std::cend(container)
+#define MOVE_ALL(container) std::make_move_iterator(std::begin(container)), std::make_move_iterator(std::end(container))
```
Comment (TarasMartynyuk) [inline SleepEngine/Engine/Utils/MacroUtils.h]: not using theese

File: SleepEngine/Engine/Components/HierarchyNode.cpp [hunk 10]
```diff
@@ -27,26 +43,32 @@ Object* HierarchyNode::addChild(std::unique_ptr<Object>&& child)
...
 
-    if (childToDelete != m_children.cend())
+void HierarchyNode::removeAllChildren()
+{
+    for (auto* const child : m_children)
     {
-        LOG_AND_FAIL("'removeChild' called for object that is not out child");
-        return;
+        slp::getHierarchy(*child).setParent(nullptr);
+        globalEntityManager().removeObjectLater(child);
```
Comment (TarasMartynyuk) [inline SleepEngine/Engine/Components/HierarchyNode.cpp]: and what about your grandchildren?

File: SleepEngine/Engine/EntityManager.cpp [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        auto const isObjectToRemove = [objectToRemove](auto const& object)
+        {
+            return objectToRemove == object.get();
+        };
+
+        auto const[found, it] = findIf(m_objects, isObjectToRemove);
+
+        if (!found)
+        {
+            LOG_AND_FAIL("EntityManager error: invalid Object pointer.");
```
Comment (TarasMartynyuk) [inline SleepEngine/Engine/EntityManager.cpp]: erro msg is so-so but ok

File: SleepEngine/Engine/EntityManager.cpp (hunk 6 above)
Comment (SleepingSoul) [reply SleepEngine/Engine/EntityManager.cpp]: In my opinion, it should be ok when few things want to destroy one other thing, and they all should be able to call this method (it should be legal)

File: SleepEngine/Engine/GameTemplate/Scene.h (hunk 8 above)
Comment (SleepingSoul) [reply SleepEngine/Engine/GameTemplate/Scene.h]: There is string -> string_view conversion so macros won't work properly)

File: SleepEngine/Engine/Utils/MacroUtils.h (hunk 9 above)
Comment (SleepingSoul) [reply SleepEngine/Engine/Utils/MacroUtils.h]: Yeah, I wanted to use this but I found a `std::move_backward` that uses simple iterators. Should I delete it? It possibly may be used in the future.

File: SleepEngine/Engine/EntityManager.cpp (hunk 11 above)
Comment (SleepingSoul) [reply SleepEngine/Engine/EntityManager.cpp]: Weird error message but ok

File: SleepEngine/Engine/Components/HierarchyNode.cpp (hunk 5 above)
Comment (SleepingSoul) [reply SleepEngine/Engine/Components/HierarchyNode.cpp]: Yes. 
When you want to completely remove the object you use `Object::removeLater()`, this is not the same thing.
This `removeChild` is being used when:
- detaching from the parent;
- changing parent to other one

File: SleepEngine/Engine/EntityManager.h (hunk 7 above)
Comment (SleepingSoul) [reply SleepEngine/Engine/EntityManager.h]: Fixed

File: SleepEngine/Engine/Components/HierarchyNode.cpp (hunk 10 above)
Comment (SleepingSoul) [reply SleepEngine/Engine/Components/HierarchyNode.cpp]: Oh shit. Fixed)

File: SleepEngine/Engine/GameTemplate/Game.cpp (hunk 2 above)
Comment (SleepingSoul) [reply SleepEngine/Engine/GameTemplate/Game.cpp]: Fixed

File: SleepEngine/Engine/EntityManager.cpp (hunk 6 above)
Comment (TarasMartynyuk) [reply SleepEngine/Engine/EntityManager.cpp]: better make this explicit some how with the ability to check if queued, but that's more complex so ok

File: SleepEngine/Engine/Components/HierarchyNode.cpp (hunk 5 above)
Comment (TarasMartynyuk) [reply SleepEngine/Engine/Components/HierarchyNode.cpp]: but erase + remove  would delete only pointer from m_children. As it is, nothing is removed, yes?

Comment (SleepingSoul) [reply SleepEngine/Engine/Components/HierarchyNode.cpp]: Not from children. Please, look at `Object::removeLater` implementation

File: SleepEngine/Engine/GameTemplate/Game.cpp (hunk 3 above)
Comment (SleepingSoul) [reply SleepEngine/Engine/GameTemplate/Game.cpp]: Fixed

**Pull request merged**